### PR TITLE
fix: Filter patches to drop `"remove"` operation to prevent dropping unknown fields

### PR DESCRIPTION
**Pull request by aramase**
**Reason for Change**:

Filter patches to only allow "add" operation to prevent dropping unknown fields

**Issue Fixed**:

fixes https://github.com/Azure/azure-workload-identity/issues/1312

### CODE REVIEW

File: pkg/webhook/webhook.go [hunk 1]
```diff
@@ -166,7 +167,30 @@ func (m *podMutator) Handle(ctx context.Context, req admission.Request) (respons
...
 	}
-	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledPod)
+
+	// patchResponse := admission.PatchResponseFromRaw(req.Object.Raw, marshaledPod)
+	// patches := []jsonpatch.JsonPatchOperation{}
+	// for _, patch := range patchResponse.Patches {
+	// 	// drop patches that are removing fields
+	// 	// this prevents patches that remove fields that the webhook doesn't know about
+	// 	// ref: https://github.com/Azure/azure-workload-identity/issues/1312
+	// 	if patch.Operation == "remove" {
```
Comment (enj) [inline pkg/webhook/webhook.go:177]: This commented out approach seems safer?  And maybe invert the check so we only keep the type of operations we understand?

File: pkg/webhook/webhook.go [hunk 2]
```diff
@@ -166,7 +167,30 @@ func (m *podMutator) Handle(ctx context.Context, req admission.Request) (respons
...
+	// 		continue
+	// 	}
+	// 	patches = append(patches, patch)
+	// }
+	//
+	// patchResponse.Patches = patches
+	// return patchResponse
+
+	// Merge the original raw object with the modified marshaled pod
+	patchedObject, err := strategicpatch.StrategicMergePatch(req.Object.Raw, marshaledPod, corev1.Pod{})
```
Comment (enj) [inline pkg/webhook/webhook.go:187]: I haven't seen any other webhook use this approach, so it makes me nervous that it won't work correctly in all cases.

File: pkg/webhook/webhook_test.go [hunk 3]
```diff
@@ -728,39 +728,44 @@ func TestHandle(t *testing.T) {
...
-			podLabels: map[string]string{
-				UseWorkloadIdentityLabel: "true",
-			},
+			name:          "pod has the required label, no warnings",
+			rawPod:        newPodRaw("pod", "ns1", "sa", map[string]string{UseWorkloadIdentityLabel: "true"}, nil, false),
+			clientObjects: serviceAccounts,
+			readerObjects: nil,
+		},
+		{
+			name:          "pod has the required label, restart policy in init container",
```
Comment (enj) [inline pkg/webhook/webhook_test.go:767]: Also add tests where we fabricate fields that will never exist at different nesting levels that don't exist anywhere in `corev1.Pod`.  Like `pod.Panda` or `pod.Spec.Panda` or `pod.Spec.Jungle.Panda`.

**Comment by aramase** [comment]
Closing this PR after further testing.

In e2e tests we create a pod with init container and test injecting sidecar proxy. The sidecar is added as the first container in the list so it's started first and the way patches get generated instead of doing a complete replace of
`/spec/containers/0`, it does replace and remove on individual fields (which is expected)
 
```yaml
{replace /spec/containers/0/name azwi-proxy}
{replace /spec/containers/0/image aramase/proxy:add_op_00}
{add /spec/containers/0/ports [map[containerPort:8080]]}
{add /spec/containers/0/securityContext/privileged false}
{replace /spec/containers/0/securityContext/runAsNonRoot true}
{add /spec/containers/0/securityContext/readOnlyRootFilesystem true}
{remove /spec/containers/0/securityContext/seccompProfile <nil>}
{remove /spec/containers/0/securityContext/runAsUser <nil>}
{remove /spec/containers/0/args/2 <nil>}
{replace /spec/containers/0/args/1 --log-level=info}
{replace /spec/containers/0/args/0 --proxy-port=8080}
```

if we skip "remove" operation, it breaks because the args and other fields incorrectly remain.

1. The k8s version bump will resolve the current issue with Kubernetes native sidecar fields.
2. I've opened  https://github.com/Azure/azure-workload-identity/issues/1337 for us to investigate further.

**Pull request closed without merging**